File: axon-frontend/src/types/photo.ts
/** Photo types. */

export interface Photo {
  id: string;
  userId: string;
  storagePath: string;
  thumbnailPath: string | null;
  fileName: string | null;
  fileSize: number | null;
  width: number | null;
  height: number | null;
  caption: string | null;
  metadata: Record<string, unknown>;
  createdAt: string;
  thumbnailUrl: string | null;
  imageUrl: string | null;
}

export interface PhotoListResponse {
  photos: Photo[];
  total: number;
  page: number;
  pageSize: number;
  hasMore: boolean;
}

export interface PhotoShareResponse {
  photoId: string;
  shareUrl: string;
  expiresAt: string;
}

export interface GallerySessionResponse {
  token: string;
  expiresAt: string;
  sessionUrl: string;
}

export interface GallerySessionPhotosResponse {
  photos: Photo[];
  total: number;
  page: number;
  pageSize: number;
  hasMore: boolean;
  expiresAt: string;
}

/** Map backend camelCase JSON to Photo (handles both casings). */
export function normalizePhoto(raw: Record<string, unknown>): Photo {
  return {
    id: String(raw.id),
    userId: String(raw.userId ?? raw.user_id),
    storagePath: String(raw.storagePath ?? raw.storage_path),
    thumbnailPath: (raw.thumbnailPath ?? raw.thumbnail_path ?? null) as string | null,
    fileName: (raw.fileName ?? raw.file_name ?? null) as string | null,
    fileSize: (raw.fileSize ?? raw.file_size ?? null) as number | null,
    width: (raw.width ?? null) as number | null,
    height: (raw.height ?? null) as number | null,
    caption: (raw.caption ?? null) as string | null,
    metadata: (raw.metadata as Record<string, unknown>) ?? {},
    createdAt: String(raw.createdAt ?? raw.created_at),
    thumbnailUrl: (raw.thumbnailUrl ?? raw.thumbnail_url ?? null) as string | null,
    imageUrl: (raw.imageUrl ?? raw.image_url ?? null) as string | null,
  };
}

export function normalizePhotoList(raw: Record<string, unknown>): PhotoListResponse {
  const photos = Array.isArray(raw.photos)
    ? raw.photos.map((item) => normalizePhoto(item as Record<string, unknown>))
    : [];
  return {
    photos,
    total: Number(raw.total ?? 0),
    page: Number(raw.page ?? 1),
    pageSize: Number(raw.pageSize ?? raw.page_size ?? 20),
    hasMore: Boolean(raw.hasMore ?? raw.has_more),
  };
}

File: axon-backend/app/schemas/photo.py
"""Photo schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.schemas.common import CamelModel


class PhotoResponse(CamelModel):
    """Photo information with signed URLs."""

    id: str
    user_id: str
    storage_path: str
    thumbnail_path: str | None = None
    file_name: str | None = None
    file_size: int | None = None
    width: int | None = None
    height: int | None = None
    caption: str | None = None
    metadata: dict = Field(default_factory=dict)
    created_at: datetime
    thumbnail_url: str | None = None
    image_url: str | None = None


class PhotoListResponse(CamelModel):
    """Paginated photo list."""

    photos: list[PhotoResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class PhotoShareResponse(CamelModel):
    """Photo share URL response."""

    photo_id: str
    share_url: str
    expires_at: datetime


class GallerySessionResponse(CamelModel):
    """Temporary gallery session for QR access."""

    token: str
    expires_at: datetime
    session_url: str


class GallerySessionPhotosResponse(CamelModel):
    """Public gallery session photo list."""

    photos: list[PhotoResponse]
    total: int
    page: int
    page_size: int
    has_more: bool
    expires_at: datetime
